[docs][material-ui] Revamp `Composition` guide (#43266)
mui/material-ui: docs/data/material/guides/composition/composition.md

You can find a code example with the Button and react-router-dom in [these demos](/material-ui/integrations/routing/#component-prop).

### Generic

> It's also possible to have a generic custom component which accepts any React component, and HTML elements.

Reviewer: nit: the term "HTML components" is probably fine just fine, but the React docs call these "built-in components" or "common components". See https://react.dev/reference/react-dom/components/common. In case you want to consider this:

```suggestion
It's also possible to have a generic custom component which accepts any React component, including [built-in components](https://react.dev/reference/react-dom/components/common).
```
